core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.resolver.dns.DnsQueryLifecycleObserver;
+
+/**
+ * A {@link DnsQueryLifecycleObserver} that helps capture custom dns metrics.
+ */
+final class DefaultDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
+
+    private final Counter success;
+    private final Counter failure;
+    private final Counter protocolType;
+    private final Counter dnsErrorCode;
+    private final Counter queryWritten;
+    private final Counter queryType;
+    private final DnsQuestion dnsQuestion;
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry, DnsQuestion question, MeterIdPrefix prefix) {
+        this.dnsQuestion = question;
+        success = meterRegistry.counter(prefix.name(), "result", "success", "question", question.name());
+        failure = meterRegistry.counter(prefix.name(), "result", "failure", "question", question.name());
+        dnsErrorCode = meterRegistry.counter(prefix.name(), "dns", "errorcodes", "question", question.name());
+        queryWritten = meterRegistry.counter(prefix.name(), question.name(), "written");
+        queryType = meterRegistry.counter(prefix.name(), question.type().name(), question.name());
+        protocolType = meterRegistry.counter(prefix.name(), "protocol", getProtocolType(question.type()),
+                                            "question", question.name());
+    }
+
+    private static String getProtocolType(DnsRecordType type) {
+        if (DnsRecordType.OPT.equals(type) ||
+                DnsRecordType.IXFR.equals(type) ||
+                DnsRecordType.AXFR.equals(type)) {
+            return "tcp";
+        }
```

Comment:
Does `OPT` really mean `tcp`? IIRC, even a UDP record can have it. If there's no way for observer to determine the protocol, then we can remove it from the metrics.

Reply:
You are correct. I think if the packet size is greater than 512 bytes then it will switch to TCP. Don't know if I can figure this inside of the observer. The observer has access to DnsQuestion but that does not give me enough information.